### PR TITLE
Hana query userstore

### PR DESCRIPTION
**Pull request by rainerleber**
##### SUMMARY
This PR add the ability to use the userstore instead of username and password

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
hana_query

##### ADDITIONAL INFORMATION
```
The hdbuserstore is a secretstore where the connection string and the credentials for a specific user is stored.
```

### CODE REVIEW

**Comment by rainerleber** [comment]
Hello everybody, this is a little improvement so that is possible to use object from the hdbuserstore.

File: plugins/modules/database/saphana/hana_query.py [hunk 1]
```diff
@@ -161,7 +178,10 @@ def main():
     if database is not None:
         command.extend(['-d', database])
     # -x Suppresses additional output, such as the number of selected rows in a result set.
-    command.extend(['-x', '-i', instance, '-u', user, '-p', password])
+    if userstore:
+        command.extend(['-x', '-U', user])
```
Comment (quidame) [inline plugins/modules/database/saphana/hana_query.py:182]: does it mean that `instance` is not required too when `userstore=True` ? This doesn't seem to be the case in `DOCUMENTATION`, `EXAMPLES` nor `argument_spec`.

File: plugins/modules/database/saphana/hana_query.py [hunk 2]
```diff
@@ -21,13 +21,17 @@
...
         required: true
     user:
-        description: A dedicated username. Defaults to C(SYSTEM).
+        description: A dedicated username. The user could be also in hdbuserstore. Defaults to C(SYSTEM).
         type: str
         default: SYSTEM
+    userstore:
+        description: If true the user must be in hdbuserstore.
+        type: bool
+        default: false
```
Comment (quidame) [inline plugins/modules/database/saphana/hana_query.py:30]:
```suggestion
        default: false
        version_added: 3.5.0
```

**Review by quidame — CHANGES_REQUESTED** [review]
@rainerleber thanks for your contribution !

I don't use this module, so my review on this PR is mainly about module interface.

This pull request also needs a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to-format) and integration tests for the new feature.

File: plugins/modules/database/saphana/hana_query.py (hunk 2 above)
Comment (quidame) [reply plugins/modules/database/saphana/hana_query.py:30]: 3.5.0 is scheduled for around August 10th (https://github.com/ansible-collections/community.general/issues/582). Seems fair, but maybe too short: the value is a suggestion, the field `version_added` is mandatory.

File: plugins/modules/database/saphana/hana_query.py (hunk 1 above)
Comment (rainerleber) [reply plugins/modules/database/saphana/hana_query.py:182]: the instance number is also used to get the file path

**Comment by rainerleber** [comment]
@quidame thank you for review and your input

File: plugins/modules/database/saphana/hana_query.py [hunk 3]
```diff
@@ -21,13 +22,18 @@
         type: str
         required: true
     user:
-        description: A dedicated username. Defaults to C(SYSTEM).
+        description: A dedicated username. The user could be also in hdbuserstore. Defaults to C(SYSTEM).
         type: str
         default: SYSTEM
+    userstore:
+        description: If true the user must be in hdbuserstore.
```
Comment (felixfontein) [inline plugins/modules/database/saphana/hana_query.py:29]:
```suggestion
        description: If C(true) the user must be in hdbuserstore.
```

File: plugins/modules/database/saphana/hana_query.py [hunk 4]
```diff
@@ -21,13 +22,18 @@
...
+        type: bool
+        default: false
+        version_added: 3.5.0
     password:
         description: The password to connect to the database.
         type: str
-        required: true
+        required: false
```
Comment (felixfontein) [inline plugins/modules/database/saphana/hana_query.py:36]:
```suggestion
```

File: plugins/modules/database/saphana/hana_query.py [hunk 5]
```diff
@@ -117,16 +134,18 @@ def main():
...
             sid=dict(type='str', required=True),
             instance=dict(type='str', required=True),
-            encrypted=dict(type='bool', required=False, default=False),
+            encrypted=dict(type='bool', default=False),
             host=dict(type='str', required=False),
-            user=dict(type='str', required=False, default="SYSTEM"),
-            password=dict(type='str', required=True, no_log=True),
+            user=dict(type='str', default="SYSTEM"),
+            userstore=dict(type='bool', default=False),
+            password=dict(type='str', required=False, no_log=True),
```
Comment (felixfontein) [inline plugins/modules/database/saphana/hana_query.py:141]:
```suggestion
            password=dict(type='str', no_log=True),
```

File: plugins/modules/database/saphana/hana_query.py [hunk 6]
```diff
@@ -161,7 +181,10 @@ def main():
     if database is not None:
         command.extend(['-d', database])
     # -x Suppresses additional output, such as the number of selected rows in a result set.
-    command.extend(['-x', '-i', instance, '-u', user, '-p', password])
+    if userstore:
+        command.extend(['-x', '-U', user])
+    else:
+        command.extend(['-x', '-i', instance, '-u', user, '-p', password])
```
Comment (felixfontein) [inline plugins/modules/database/saphana/hana_query.py:187]: BTW, since this is a common security problem: is there a way of providing the password in another way, say via stdin? That would be preferable over passing them on the command line, since that allows other users on the same system to extract the password from `/proc`.

**Comment by quidame** [comment]
@rainerleber , could you please add either integration tests (I would prefer, but there is none yet) or unit tests (there are already some for this module) to validate that the new option is working as expected and is failing as expected ?

File: plugins/modules/database/saphana/hana_query.py (hunk 6 above)
Comment (rainerleber) [reply plugins/modules/database/saphana/hana_query.py:187]: for this case is the userstore as an alternative to prevent the password is available anywhere

**Comment by rainerleber** [comment]
> @rainerleber , could you please add either integration tests (I would prefer, but there is none yet) or unit tests (there are already some for this module) to validate that the new option is working as expected and is failing as expected ?

Thank you for the review. Integration tests are hard to implement because this module needs a HANA database. I added two more unit tests. :-)

File: plugins/modules/database/saphana/hana_query.py [hunk 7]
```diff
@@ -21,13 +22,17 @@
...
+        description: A dedicated username. The user could be also in hdbuserstore. Defaults to C(SYSTEM).
         type: str
         default: SYSTEM
+    userstore:
+        description: If C(true) the user must be in hdbuserstore.
+        type: bool
+        default: false
+        version_added: 3.5.0
     password:
         description: The password to connect to the database.
```
Comment (felixfontein) [inline plugins/modules/database/saphana/hana_query.py:34]:
```suggestion
        description:
          - The password to connect to the database.
          - "B(Note:) Since the passwords have to be passed as command line arguments, I(userstore=true) should
            be used whenever possible, as command line arguments can be seen by other users
            on the same machine."
```

Comment (rainerleber) [reply plugins/modules/database/saphana/hana_query.py:34]: Thank you

**Review by felixfontein — APPROVED** [review]

**Review by quidame — APPROVED** [review]
LGTM

**Comment by felixfontein** [comment]
@rainerleber thanks for your contribution!
@quidame thanks for reviewing!
